designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,556 @@
+
+The latent data block appears after the transformed parameters block
+but before the odel block.  Like the parameters, transformed
+parameters, and generated quantities, the latent data will be included
+in the output and the top-level variables will be available in all
+later blocks.  Latent data will also have access to the data,
+transformed data, parameters, and transformed parameters.
+
+#### Iteration number 
+
+The iteration number can be optionally set in the latent data block 
+from the outside.  There will be an `iteration_number__` parameter 
+reserved.  It will be set to 0 by default, but can be updated by an 
+external algorithm. 
+
+### Program Execution for latent data
+
+For all of our inference algorithms (HMC, ADVI, Pathfinder,
+optimization, Laplace), the latent data block should be executed once
+at the start of each iteration.  For example, in Hamiltonian Monte
+Carlo, the latent data block is executed once based on the initial
+parameter values, then the values are used without changing each
+leapfrog iteration.  For optimization, each iteration of L-BFGS should
+evaluate the latent data block once per iteration and leave it fixed
+through the line search. The latent data needs to be executed before
+the model block.
+
+By default, the latent data block will be empty and there will be
+nothing to compute.  This should be possible to set up so that all of
+the inference methods are backwards compatible.
+
+As with blocks other than the parameters block, any constraints on the
+declared variables will be checked at the end of the block and an
+exception will be raised if they are violated, which will cause the
+current iteration in any of the algorithms to be rejected.
+
+## C++ model class  
+
+The `latent data` block will be executed with primitive (`double` and
+`int` in C++) variables, just like the `generated quantities` block.
+This effectively "cuts" the gradient information from propagating back
+through the latent data block.  Nevertheless, computation in the
+latent data block can affect what happens in the model block by
+defining new variables on which the model block may condition.
+
+### Representing the latent data
+
+The design is based on an object-based representation of the latent
+data using code generation.  For example, suppose the latent data
+block is defined as follows.
+
+
+```cpp
+latent data {
+  real<lower=0, upper=1> sens = beta_rng(98, 2);
+  real<lower=0, upper=1> spec = beta_rng(95, 5);
+}
+```
+
+This will lead to the code generation of a class in the generated
+`.hpp` file for the model.
+
+```cpp
+struct LatentData {
+  int iteration_number__;  // included by default in all LatentData
+  double sens;
+  double spec;
+};
+```
+
+### Generating the latent data
+
+An additional method in the model class will be required to populate 
+the latent data. 
+
+```cpp
+template <typename RNG> inline void
+latent_data(
+    const VectorXd& params_r,
+    LatentData& latent_data__,
```

Comment:
Two items of concern here, but they're detail-level:

1. The model base class will not be able to actually name the LatentData, will it? I'm not 100% on the rules for forward-declaring, but if I'm correct, this will need to be a void pointer or something instead
2. I believe that also would mean this has a chicken-egg problem because it's not clear how to create the first iteration's LatentData

@SteveBronder and I discussed a few ways to implement this. I think a simple one is `latent_data_size()` which returns a size_t, and then a `void* latent_data` argument to this function which is passed a pointer created by `malloc(latent_data_size())`. I think he had a more C++-idiomatic solution that was conceptually similar

Reply:
A (nullable) pointer may also be preferable when actually calling log_prob, because that would let log_prob check if it was called with null, and throw an error if the model actually contained latent data. The default-constructed overload you suggest wouldn't allow such a check, and I think that could lead to annoyances in downstream things like bridgestan

Reply:
1. **Generic handling of `LatentData` type from base class.**  We can have it return a pointer that gets cast by the actual model.  This is what we do elsewhere for the model class itself with a pointer-based factory method and I believe we can just follow that pattern.  I'll update the proposal.

2. **First iteration**.  I don't think there's an issue with the first iteration because the latent data is generated based on the parameters and data, and those will be set for the first iteration.  Or did you just mean to construct it initially?   For that, we'll need the generic higher-level factory method call.

3.  **Null argument to `log_prob`**:   With the above design it'll almost have to be a null pointer argument which can be cast internally if non-null.

P.S.  I wasn't expecting this to be approved as is---I just wanted to put it up in a way that would be easy to discuss.

Reply:
I'm marking this as resolved with Steve's suggestion, which I'm incorporating.